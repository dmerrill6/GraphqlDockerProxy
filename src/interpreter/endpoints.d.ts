export interface Endpoint {
  url: string;
  namespace: string;
  typePrefix: string;
  __created: string;
  __intospection?: object;
  __imageID: string;
  __burnd?: boolean;
  __deploymentName : string;
  __loadbalance?: __loadbalance;
}

interface __loadbalance {
  count: number;
  endpoints: Endpoints;
}

export type Endpoints = {
  [key : string]: Endpoint[],
};