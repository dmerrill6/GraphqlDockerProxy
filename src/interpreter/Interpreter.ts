import { runtime } from '../properties'
export class Interpreter {

  resetConnection = () => {
    winston.info('Reset Connection is not implemented for ' + runtime)
  }
}
