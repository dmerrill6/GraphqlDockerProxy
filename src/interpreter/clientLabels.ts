export const TOKEN:string = 'gqlProxy.token'
export const URL:string = 'gqlProxy.url'
export const NAMESPACE:string = 'gqlProxy.namespace'
