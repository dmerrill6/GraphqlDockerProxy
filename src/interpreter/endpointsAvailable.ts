import fetch from 'node-fetch'
import  { Endpoints } from './endpoints'
// tslint:disable-next-line
const queryStr = '?query=%0A%20%20%20%20query%20IntrospectionQuery%20%7B%0A%20%20%20%20%20%20__schema%20%7B%0A%20%20%20%20%20%20%20%20queryType%20%7B%20name%20%7D%0A%20%20%20%20%20%20%20%20mutationType%20%7B%20name%20%7D%0A%20%20%20%20%20%20%20%20subscriptionType%20%7B%20name%20%7D%0A%20%20%20%20%20%20%20%20types%20%7B%0A%20%20%20%20%20%20%20%20%20%20...FullType%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20directives%20%7B%0A%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20description%0A%20%20%20%20%20%20%20%20%20%20locations%0A%20%20%20%20%20%20%20%20%20%20args%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20...InputValue%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%0A%20%20%20%20fragment%20FullType%20on%20__Type%20%7B%0A%20%20%20%20%20%20kind%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20description%0A%20%20%20%20%20%20fields(includeDeprecated%3A%20true)%20%7B%0A%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20description%0A%20%20%20%20%20%20%20%20args%20%7B%0A%20%20%20%20%20%20%20%20%20%20...InputValue%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20type%20%7B%0A%20%20%20%20%20%20%20%20%20%20...TypeRef%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20isDeprecated%0A%20%20%20%20%20%20%20%20deprecationReason%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20inputFields%20%7B%0A%20%20%20%20%20%20%20%20...InputValue%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20interfaces%20%7B%0A%20%20%20%20%20%20%20%20...TypeRef%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20enumValues(includeDeprecated%3A%20true)%20%7B%0A%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20description%0A%20%20%20%20%20%20%20%20isDeprecated%0A%20%20%20%20%20%20%20%20deprecationReason%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20possibleTypes%20%7B%0A%20%20%20%20%20%20%20%20...TypeRef%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%0A%20%20%20%20fragment%20InputValue%20on%20__InputValue%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20description%0A%20%20%20%20%20%20type%20%7B%20...TypeRef%20%7D%0A%20%20%20%20%20%20defaultValue%0A%20%20%20%20%7D%0A%0A%20%20%20%20fragment%20TypeRef%20on%20__Type%20%7B%0A%20%20%20%20%20%20kind%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20' 
export const sortEndpointAndFindAvailableEndpoints =
async(endpoints :Endpoints): Promise<Endpoints> => {
  for (const namespace in endpoints) {
    const oneNamespace = endpoints[namespace]
    for (let i = 0; i < oneNamespace.length; i = i + 1) {
      const one = oneNamespace[i]
      const result = await fetch(one.url + queryStr, {
        timeout: 2000,
      }).then(async(res) => {
        return {
          json: await res.json(),
          status: res.status,
          ok: res.ok,
        }
      })
        .catch((err) => {
          winston.debug('Error by fetching', { err, url: one.url })
          return {
            json: {},
            status: 404,
            ok: false,
          }
        })
      if (!result.ok) {
        winston.warn('Endpoint not Available: ', { url: one.url , namespace: one.namespace })
        oneNamespace.splice(i, 1)
      } else {
        oneNamespace[i].__introspection = result.json
      }

    }
    if (oneNamespace.length === 0) {
      delete endpoints[namespace]
    }

  }
  return endpoints
}
