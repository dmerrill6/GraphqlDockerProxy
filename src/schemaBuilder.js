import { HttpLink } from 'apollo-link-http'
import { setContext } from 'apollo-link-context'
import { makeRemoteExecutableSchema, mergeSchemas, introspectSchema } from 'graphql-tools'
import fetch from 'node-fetch'
import type { Endpoint } from './finder/findEndpointsInterface'


export const createRemoteSchema = async(url) => {
  const http = new HttpLink({ uri: url, fetch })
  const link = setContext((request, previousContext) => {
    return previousContext.graphqlContext
  }).concat(http)

  const schema = await introspectSchema(link)
  const executableSchema = makeRemoteExecutableSchema({
    schema,
    link,
  })
  return executableSchema
}

export const getMergedInformation = async(namespace: Array<Endpoint>) => {
  const schema = []

  for (let i = 0; i < namespace.length; i++){
    schema.push(await createRemoteSchema(namespace[i].url))
  }

  const merged = mergeSchemas({
    schemas: schema,
  })
  return merged
}
