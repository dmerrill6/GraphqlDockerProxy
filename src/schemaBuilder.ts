import { HttpLink } from 'apollo-link-http'
import { setContext } from 'apollo-link-context'
import { makeRemoteExecutableSchema, mergeSchemas, introspectSchema } from 'graphql-tools'
import fetch from 'node-fetch'
import  { Endpoint } from './interpreter/endpoints'

export const createRemoteSchema = async(url : string) => {
  const http = new HttpLink({ fetch, uri: url  })
  const link = setContext((request, previousContext) => {
    return previousContext.graphqlContext
  }).concat(http)

  const schema = await introspectSchema(link)
  const executableSchema = makeRemoteExecutableSchema({
    schema,
    link,
  })
  return executableSchema
}

export const getMergedInformation = async(namespace: Endpoint[]) => {
  const schema = []

  for (let i = 0; i < namespace.length; i = i + 1) {
    schema.push(await createRemoteSchema(namespace[i].url))
  }

  const merged = mergeSchemas({
    schemas: schema,
  })
  return merged
}
