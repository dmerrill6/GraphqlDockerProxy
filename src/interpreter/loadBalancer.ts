import * as express from 'express';
import * as request from 'request';
import { sortEndpointAndFindAvailableEndpoints } from './endpointsAvailable';
import { Endpoints } from './endpoints';
let lBserver = [];

export const foundEquals = async(datas: Endpoints) :Promise<Endpoints> => {
  const data = await sortEndpointAndFindAvailableEndpoints(datas);
  closeAllServer();
  for (const one in data) {
    const namespace = data[one];
    for (let i = 0, l = namespace.length; i < l; i = i + 1) {
      const lbData = {};
      const searchingElement = namespace[i];
      if (searchingElement.__burnd === undefined) {
        for (let j = i + 1; j < l; j = j + 1) {
          const testingElement = namespace[j];

          if (searchingElement.__imageID === testingElement.__imageID) {
            if (lbData[searchingElement.__imageID] === undefined) {
              lbData[searchingElement.__imageID] = [];
            }
            lbData[searchingElement.__imageID].push(testingElement);
            namespace[j].__burnd = true;
          // namespace.splice(j)
          }

        }
        if (lbData[searchingElement.__imageID] !== undefined) {
          lbData[searchingElement.__imageID].push(searchingElement);
          const lbResult = loadANewLoadBalaceMiddleware(lbData[searchingElement.__imageID]);
          namespace[i].url = lbResult.url;
          namespace[i].__loadbalance = {
            count: namespace.length,
            endpoints: lbResult.clients,
          };

        }
      }
      const newNamespaces = [];
      for (const oneBurable in namespace) {
        if (!namespace[oneBurable].__burnd) {
          newNamespaces.push(namespace[oneBurable]);
        }
      }
      data[one] = newNamespaces;

    }

  }

  return data;
};

export const loadANewLoadBalaceMiddleware = (listOfBackends) => {

  const servers = listOfBackends.map((one) => {
    return one.url;
  });
  let cur = 0;

  const handler = (req, res) => {

    // console.log('hier:', req.headers, servers.length)
    if (req.url === '/') {
      req.url = '';
    }
    req.pipe(request({ url: servers[cur] + req.url })).pipe(res);
    cur = (cur + 1) % servers.length;
  };
  const server = express();
  server.get('*', handler).post('*', handler);
  server.post('*', handler).post('*', handler);
  const port = getNextPortNumber();
  lBserver.push(server.listen(port));
  return {
    url: 'http://127.0.0.1:' + port + '',
    clients: listOfBackends.map((one) => {
      return Object.assign({}, one);
    }),
  };
};

let nextPortNumber = 0;
const getNextPortNumber = () => {
  nextPortNumber = nextPortNumber + 1;
  return 8000 + nextPortNumber;
};

export const closeAllServer = () => {
  lBserver.forEach((one) => {
    winston.info('close load balancer', one._connectionKey);
    one.close();
  });
  nextPortNumber = 0;
  lBserver = [];
};
