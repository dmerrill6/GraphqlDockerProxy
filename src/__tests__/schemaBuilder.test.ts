import { createRemoteSchema, getMergedInformation } from '../schemaBuilder'
import { Endpoints } from '../interpreter/endpoints'
jest.mock('graphql-tools', () => {
  return {
    introspectSchema: (link) => {
      return Promise.resolve(link)
    },

    makeRemoteExecutableSchema: (obj) => {
      return obj
    },
    mergeSchemas: (obj) => {
      return obj
    },
  }
})

jest.mock('apollo-link-context', () => {
  return {
    setContext: () => {
      return {
        concat: (link) => {
          return link
        },
      }
    },
  }
})
jest.mock('apollo-link-http', () => {
  return {
    HttpLink: class {
      uri: string
      constructor({ uri }) {
        this.uri = uri
      }
    },
  }
})

describe('test schemabuilder', () => {
  it('tests createRemoteSchema', async() => {
    expect(await createRemoteSchema('mock')).toMatchSnapshot()
  })

  it('tests getMergedInformation', async() => {
    const endpoints: Endpoints = {
      a: [
        {
          url: 'mock',
          namespace: 'mock',
          __deploymentName: '---',
          __imageID: '---',
          typePrefix: '_mock',
        },
      ],
    }

    expect(await getMergedInformation(endpoints.a)).toMatchSnapshot()
  })
})
