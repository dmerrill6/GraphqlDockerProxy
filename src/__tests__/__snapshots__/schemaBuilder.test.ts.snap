// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test schemabuilder tests createRemoteSchema 1`] = `
Object {
  "link": class_1 {
    "uri": "mock",
  },
  "schema": class_1 {
    "uri": "mock",
  },
}
`;

exports[`test schemabuilder tests getMergedInformation 1`] = `
Object {
  "schemas": Array [
    Object {
      "link": class_1 {
        "uri": "mock",
      },
      "schema": class_1 {
        "uri": "mock",
      },
    },
  ],
}
`;
